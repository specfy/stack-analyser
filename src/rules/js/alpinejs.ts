import { register } from '../../register.js';

register({
  tech: 'alpinejs',
  name: 'Alpine.js',
  type: 'tool',
  dependencies: [{ type: 'npm', name: 'alpinejs' }],
});
