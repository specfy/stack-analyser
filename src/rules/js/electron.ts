import { register } from '../../register.js';

register({
  tech: 'electron',
  name: 'Electron',
  type: 'tool',
  dependencies: [{ type: 'npm', name: 'electron' }],
});
