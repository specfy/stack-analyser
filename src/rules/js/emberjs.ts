import { register } from '../../register.js';

register({
  tech: 'emberjs',
  name: 'Ember',
  type: 'tool',
  dependencies: [
    { type: 'npm', name: 'ember-cli' },
    { type: 'npm', name: 'ember' },
    { type: 'npm', name: 'ember-data' },
    { type: 'npm', name: 'ember-source' },
  ],
});
