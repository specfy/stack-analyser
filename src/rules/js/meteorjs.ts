import { register } from '../../register.js';

register({
  tech: 'meteorjs',
  name: 'Meteor',
  type: 'tool',
  files: ['.meteor'],
  dependencies: [
    { type: 'npm', name: 'meteor' },
    { type: 'npm', name: 'meteor-node-stubs' },
  ],
});
