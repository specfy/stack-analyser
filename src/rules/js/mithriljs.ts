import { register } from '../../register.js';

register({
  tech: 'mithriljs',
  name: 'Mithril',
  type: 'tool',
  dependencies: [{ type: 'npm', name: 'mithril' }],
});
