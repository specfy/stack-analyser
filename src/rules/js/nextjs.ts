import { register } from '../../register.js';

register({
  tech: 'nextjs',
  name: 'Next.js',
  type: 'tool',
  files: ['next.config.js'],
  dependencies: [{ type: 'npm', name: 'next' }],
});
