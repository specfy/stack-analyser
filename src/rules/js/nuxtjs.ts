import { register } from '../../register.js';

register({
  tech: 'nuxtjs',
  name: 'Nuxt.js',
  type: 'tool',
  files: ['nuxt.config.js'],
  dependencies: [{ type: 'npm', name: 'nuxt' }],
});
