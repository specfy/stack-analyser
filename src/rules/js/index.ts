import './alpinejs.js';
import './angular.js';
import './bootstrap.js';
import './emberjs.js';
import './esbuild.js';
import './eslint.js';
import './express.js';
import './fastify.js';
import './koa.js';
import './meteorjs.js';
import './mithriljs.js';
import './nextjs.js';
import './nuxtjs.js';
import './preactjs.js';
import './prettier.js';
import './prisma.js';
import './puppeteer.js';
import './react.js';
import './reactEmail.js';
import './rollup.js';
import './sequelize.js';
import './socketio.js';
import './solidjs.js';
import './storybook.js';
import './strapi.js';
import './stylelint.js';
import './sveltejs.js';
import './tailwind.js';
import './typescript.js';
import './vite.js';
import './vue.js';
import './webpack.js';
