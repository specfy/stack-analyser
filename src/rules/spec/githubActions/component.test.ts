import { describe, expect, it } from 'vitest';

import { analyser } from '../../../analyser/index.js';
import { rawList } from '../../../loader.js';
import { flatten } from '../../../payload/helpers.js';
import { FakeProvider } from '../../../provider/fake.js';

import '../../../autoload.js';
import { detectGithubActionsComponent } from './component.js';

describe('docker', () => {
  it('should match all dependencies', async () => {
    const yml: string[] = [
      `name: Main
on:
  push:
    branches: [ main ]

jobs:
  ci:
    container: node:18.17.1
    runs-on: ubuntu-latest
    timeout-minutes: 15
    services:
      postgres:
        env:
          POSTGRES_PASSWORD: postgres
        image: postgres:15.4-alpine
        options: >-
          --health-cmd pg_isready
          --health-interval 10s

    env:
      OPENAI_KEY: ""

    steps:
      - name: Check out repository code
        run: npm test`,
    ];
    for (const item of rawList) {
      if (item.type !== 'dependency' || item.ref.type !== 'githubAction') {
        continue;
      }

      const example =
        'example' in item.ref ? item.ref.example : `${item.ref.name}@1`;
      const [name] = example.split('@');
      yml.push(`
      - name: ${name}
        uses: ${example}
    `);
    }

    const res = await analyser({
      provider: new FakeProvider({
        paths: {
          '/': ['.github/'],
          '/.github': ['workflows/'],
          '/.github/workflows': ['main.yml'],
        },
        files: {
          '/.github/workflows/main.yml': yml.join(''),
        },
      }),
    });

    const merged = flatten(res, { merge: true });
    expect(Array.from(merged.techs).sort()).toMatchSnapshot();
    expect(Array.from(merged.dependencies).sort()).toMatchSnapshot();
  });

  it('should gracefully fail to parse', async () => {
    const res = await detectGithubActionsComponent(
      [{ fp: '/.github/workflows/main.yml', name: 'main.yml', type: 'file' }],
      new FakeProvider({
        paths: {
          '/': ['.github/'],
          '/.github': ['workflows/'],
          '/.github/workflows': ['main.yml'],
        },
        files: {
          '/.github/workflows/main.yml': '',
        },
      })
    );

    expect(res).toBe(false);
  });
});
