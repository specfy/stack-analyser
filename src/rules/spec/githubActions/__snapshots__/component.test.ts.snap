// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`docker > should match all dependencies 1`] = `
[
  "algolia",
  "alibabacloud",
  "aws",
  "aws.ec2",
  "aws.fargate",
  "aws.lambda",
  "aws.s3",
  "azure",
  "databricks",
  "datadog",
  "digitalocean",
  "discord",
  "expodev",
  "firebase",
  "flyio",
  "gcp",
  "gcp.cloudrun",
  "gcp.functions",
  "gcp.gce",
  "gcp.gcs",
  "gcp.gke",
  "gcp.secretmanager",
  "github",
  "github.actions",
  "github.pages",
  "hashicorp_vault",
  "heroku",
  "kubernetes",
  "launchdarkly",
  "netlify",
  "planetscale",
  "postgresql",
  "scaleway",
  "sentry",
  "slack",
  "snyk",
  "supabase",
  "supabase.postgres",
  "twilio",
  "vercel",
]
`;

exports[`docker > should match all dependencies 2`] = `
[
  [
    "docker",
    "node",
    "18.17.1",
  ],
  [
    "docker",
    "postgres",
    "15.4-alpine",
  ],
  [
    "githubAction",
    "10up/action-wordpress-plugin-deploy",
    "1",
  ],
  [
    "githubAction",
    "8398a7/action-slack",
    "1",
  ],
  [
    "githubAction",
    "CircleCI-Public/trigger-circleci-pipeline-action",
    "1",
  ],
  [
    "githubAction",
    "DataDog/synthetics-ci-github-action",
    "1",
  ],
  [
    "githubAction",
    "Ilshidur/action-discord",
    "1",
  ],
  [
    "githubAction",
    "JamesIves/github-pages-deploy-action",
    "1",
  ],
  [
    "githubAction",
    "actions/deploy-pages",
    "1",
  ],
  [
    "githubAction",
    "actions/upload-pages-artifact",
    "1",
  ],
  [
    "githubAction",
    "ahmadnassri/action-dependabot-auto-merge",
    "1",
  ],
  [
    "githubAction",
    "akhileshns/heroku-deploy",
    "1",
  ],
  [
    "githubAction",
    "algolia/algoliasearch-crawler-github-actions",
    "1",
  ],
  [
    "githubAction",
    "algolia/setup-algolia-cli",
    "1",
  ],
  [
    "githubAction",
    "aliyun/acr-login",
    "1",
  ],
  [
    "githubAction",
    "amondnet/vercel-action",
    "1",
  ],
  [
    "githubAction",
    "appleboy/lambda-action",
    "1",
  ],
  [
    "githubAction",
    "aws-actions/amazon-eks-fargate",
    "1",
  ],
  [
    "githubAction",
    "aws-actions/configure-aws-credentials",
    "1",
  ],
  [
    "githubAction",
    "azure/CLI",
    "1",
  ],
  [
    "githubAction",
    "azure/k8s-create-secret",
    "1",
  ],
  [
    "githubAction",
    "azure/k8s-deploy",
    "1",
  ],
  [
    "githubAction",
    "azure/k8s-set-context",
    "1",
  ],
  [
    "githubAction",
    "azure/login",
    "1",
  ],
  [
    "githubAction",
    "azure/powershell",
    "1",
  ],
  [
    "githubAction",
    "azure/setup-helm",
    "1",
  ],
  [
    "githubAction",
    "azure/setup-kubectl",
    "1",
  ],
  [
    "githubAction",
    "bitovi/github-actions-deploy-docker-to-ec2",
    "1",
  ],
  [
    "githubAction",
    "cypress-io/github-action",
    "1",
  ],
  [
    "githubAction",
    "databricks/run-notebook",
    "1",
  ],
  [
    "githubAction",
    "datadog/agent-github-action",
    "1",
  ],
  [
    "githubAction",
    "digitalocean/action-doctl",
    "1",
  ],
  [
    "githubAction",
    "digitalocean/app_action",
    "1",
  ],
  [
    "githubAction",
    "docker/login-action",
    "1",
  ],
  [
    "githubAction",
    "expo/expo-github-action",
    "1",
  ],
  [
    "githubAction",
    "getsentry/action-release",
    "1",
  ],
  [
    "githubAction",
    "google-github-actions/auth",
    "1",
  ],
  [
    "githubAction",
    "google-github-actions/deploy-cloud-functions",
    "1",
  ],
  [
    "githubAction",
    "google-github-actions/deploy-cloudrun",
    "1",
  ],
  [
    "githubAction",
    "google-github-actions/get-gke-credentials",
    "1",
  ],
  [
    "githubAction",
    "google-github-actions/get-secretmanager-secrets",
    "1",
  ],
  [
    "githubAction",
    "google-github-actions/setup-gcloud",
    "1",
  ],
  [
    "githubAction",
    "google-github-actions/ssh-compute",
    "1",
  ],
  [
    "githubAction",
    "google-github-actions/upload-cloud-storage",
    "1",
  ],
  [
    "githubAction",
    "hashicorp/setup-terraform",
    "1",
  ],
  [
    "githubAction",
    "hashicorp/terraform-cdk-action",
    "1",
  ],
  [
    "githubAction",
    "hashicorp/vault-action",
    "1",
  ],
  [
    "githubAction",
    "jakejarvis/s3-sync-action",
    "1",
  ],
  [
    "githubAction",
    "launchdarkly/find-code-references",
    "1",
  ],
  [
    "githubAction",
    "nwtgck/actions-netlify",
    "1",
  ],
  [
    "githubAction",
    "peaceiris/actions-gh-pages",
    "1",
  ],
  [
    "githubAction",
    "planetscale/create-branch-action",
    "1",
  ],
  [
    "githubAction",
    "relative-ci/agent-action",
    "1",
  ],
  [
    "githubAction",
    "scaleway/action-scw",
    "1",
  ],
  [
    "githubAction",
    "slackapi/slack-github-action",
    "1",
  ],
  [
    "githubAction",
    "snaplet/vercel-action",
    "1",
  ],
  [
    "githubAction",
    "snyk/actions/php",
    undefined,
  ],
  [
    "githubAction",
    "supabase/embeddings-generator",
    "1",
  ],
  [
    "githubAction",
    "supabase/setup-cli",
    "1",
  ],
  [
    "githubAction",
    "superfly/flyctl-actions/setup-flyctl",
    "1",
  ],
  [
    "githubAction",
    "trufflesecurity/trufflehog",
    "1",
  ],
  [
    "githubAction",
    "twilio-labs/actions-sms",
    "1",
  ],
  [
    "githubAction",
    "w9jds/firebase-action",
    "1",
  ],
  [
    "githubAction",
    "wzieba/Firebase-Distribution-Github-Action",
    "1",
  ],
]
`;
